File: server/card.php
<?php

class Card {
    public string $name;
    public array $narration;
    public Closure $action;
    public Closure $conditions;

    public function __construct(string $name, callable $conditions, callable $action) {
        $this->name = $name;
        $this->conditions = $conditions(...); 
        $this->action = $action(...); 
    }

    public function is_playable(array &$state) {
        return ($this->conditions)($state);
    }

    public function play(array &$state): string | null {
        if (!$this->is_playable($state)) { return null; }
        return ($this->action)($state);
    }
}

function playable_cards(array $deck, array $state): array {
    $default_card = [
        'skip' => new Card(
            'skip',
            function() { return true; },
            function(&$state) {
                clear_event_logs($state);
                end_turn($state);
                log_event($state, "Let's see what else is happening.");
            }
        ),
    ];
    $playable_cards = array_filter($deck, function ($card) use($state) { return $card->is_playable($state); });
    return count($playable_cards) > 0 ? $playable_cards : $default_card;
}

File: server-go/card.go
package main

type Card struct {
	Name       string
	Conditions func(state *GameState) bool
	Action     func(state *GameState)
}

func PlayableCards(deck map[string]*Card, state *GameState) map[string]*Card {
	playable := make(map[string]*Card)
	for key, card := range deck {
		if card.Conditions(state) {
			playable[key] = card
		}
	}

	if len(playable) == 0 {
		return map[string]*Card{"skip": deck["skip"]}
	}
	return playable
}

func ChooseCard(p *Player, playable map[string]*Card, state *GameState) *Card {
	key := p.Decide(state)
	if card, ok := playable[key]; ok {
		return card
	}
	return playable["skip"]
}